docs/modules/architecture/pages/compact-binary-specification.adoc

```diff
@@ -0,0 +1,706 @@
+
+In the Schema class, each field will either
+
+* have a positive offset, if it is a fixed-size field
+* have a positive index if it is a variable-size field
+* have a positive bit offset if it is a boolean field. This bit offset is the offset within the byte given by the normal offset
+
+== Schema ID
+
+We are using 64bit https://en.wikipedia.org/wiki/Rabin_fingerprint[Rabin fingerprint] to create [.inline-comment-marker]#a schema id#.
+
+Rabin fingerprint is chosen mostly because it is recommended in Avro’s documents as follows +
+https://avro.apache.org/docs/current/spec.html#schema_primitive
+
+[source,text]
+----
+At the opposite extreme, the smallest fingerprint recommended is a 64-bit Rabin fingerprint. Below, there is a provided pseudo-code for this algorithm that can be easily translated into any programming language. 64-bit fingerprints should guarantee uniqueness for schema caches of up to a million entries (for such a cache, the chance of a collision is 3E-8). It is not recommended to use shorter fingerprints, as the chances of collisions are too high (for example, with 32-bit fingerprints, a cache with as few as 100,000 schemas has a 50% chance of having a collision).
+----
+
+The schema id is calculated from the byte array representation of the schema described above.
+
+The implementation that is used is as follows:
+
+[source,java,linenums]
+----
+long fingerprint64(byte[] buf) {
+  if (FP_TABLE == null) initFPTable();
+  long fp = EMPTY;
+  for (int i = 0; i < buf.length; i++)
+    fp = (fp >>> 8) ^ FP_TABLE[(int)(fp ^ buf[i]) & 0xff];
+  return fp;
+}
+
+static long EMPTY = 0xc15d213aa4d7a795L;
+static long[] FP_TABLE = null;
+
+void initFPTable() {
+  FP_TABLE = new long[256];
+  for (int i = 0; i < 256; i++) {
+    long fp = i;
+    for (int j = 0; j < 8; j++)
+      fp = (fp >>> 1) ^ (EMPTY & -(fp & 1L));
+    FP_TABLE[i] = fp;
+  }
+}
+----
+
+== Arrays
+
+Arrays of fix-sized items can not have `null` items. On the other hand, arrays of variable-size items may contain `null` items.
+
+=== Array of Fixed-size Items
+
+[cols=",",options="header",]
+|===
+|Name |Type
+|Number of items |i32
+|item 0 |item type
+|item 1 |item type
+|item 2 |item type
+|item n |item type
+|===
+
+=== Array of Variable-size Items
+
+Consists of `Header`, `Data`, and `Offsets` sections in this order.
+
+=== Header Section
+
+[cols="1,1"]
+|===
+|Name |Type
+|Data length |i32
+|Number of items |i32
+|===
+
+=== Data Section
+
+[cols="1,1"]
+|===
+|Name |Type
+|Item 0 | item type
+|Item 1 | item type
+|... | ...
+|Item n | item type
+|===
+
+=== Offsets Section
+
+[cols="1,1"]
+|===
+|Name |Type
+|Item 0 offset | u8/u16/i32
+|Item 1 offset | u8/u16/i32
+|... | ...
+|Item n offset | u8/u16/i32
+|===
+
+
+---
+
+An array can contain only a single type of item.
+In the case of `Compact[]` all the items must have the same schema, i.e their schema id must be equal. 
```

Comment:
In the case of `Compact[]`, all the items must have the same schema, i.e, their schema ID must be equal.